File: Raven.Client.Lightweight/Document/Batches/LazyStartsWithOperation.cs
// -----------------------------------------------------------------------
//  <copyright file="LazyStartsWithOperation.cs" company="Hibernating Rhinos LTD">
//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------
namespace Raven.Client.Document.Batches
{
	using System;
	using System.Collections.Generic;
	using System.Linq;

	using Raven.Abstractions.Data;
	using Raven.Client.Connection;
	using Raven.Client.Extensions;
	using Raven.Client.Shard;
	using Raven.Json.Linq;

	public class LazyStartsWithOperation<T> : ILazyOperation
	{
		private readonly string keyPrefix;

		private readonly string matches;

	    private readonly string exclude;

		private readonly int start;

		private readonly int pageSize;

		private readonly InMemoryDocumentSessionOperations sessionOperations;

	    private readonly RavenPagingInformation pagingInformation;

	    public LazyStartsWithOperation(string keyPrefix, string matches, string exclude, int start, int pageSize, InMemoryDocumentSessionOperations sessionOperations, RavenPagingInformation pagingInformation)
		{
			this.keyPrefix = keyPrefix;
			this.matches = matches;
		    this.exclude = exclude;
			this.start = start;
			this.pageSize = pageSize;
			this.sessionOperations = sessionOperations;
		    this.pagingInformation = pagingInformation;
		}

		public GetRequest CreateRequest()
		{
            var actualStart = start;

            var nextPage = pagingInformation != null && pagingInformation.IsForPreviousPage(start, pageSize);
            if (nextPage)
                actualStart = pagingInformation.NextPageStart;

			return new GetRequest
			{
				Url = "/docs",
				Query =
					string.Format(
						"startsWith={0}&matches={3}&exclude={4}&start={1}&pageSize={2}&next-page={5}",
						Uri.EscapeDataString(keyPrefix),
                        actualStart.ToInvariantString(),
						pageSize.ToInvariantString(),
						Uri.EscapeDataString(matches ?? ""),
                        Uri.EscapeDataString(exclude ?? ""),
                        nextPage ? "true" : "false")
			};
		}

		public object Result { get; set; }

		public QueryResult QueryResult { get; set; }

		public bool RequiresRetry { get; set; }

		public void HandleResponse(GetResponse response)
		{
			if (response.RequestHasErrors())
			{
				Result = null;
				RequiresRetry = false;
				return;
			}

			var jsonDocuments = SerializationHelper.RavenJObjectsToJsonDocuments(((RavenJArray)response.Result).OfType<RavenJObject>());

            int nextPageStart;
            if (pagingInformation != null && int.TryParse(response.Headers[Constants.NextPageStart], out nextPageStart))
                pagingInformation.Fill(start, pageSize, nextPageStart);

			Result = jsonDocuments
				.Select(sessionOperations.TrackEntity<T>)
				.ToArray();
		}

        public void HandleResponses(GetResponse[] responses, ShardStrategy shardStrategy)
		{
			if (responses.Any(x => x.RequestHasErrors()))
			{
				Result = null;
				RequiresRetry = true;
				return;
			}

			var jsonDocuments = new List<JsonDocument>();
			foreach (var response in responses)
			{
				var documents = SerializationHelper.RavenJObjectsToJsonDocuments(((RavenJArray)response.Result).OfType<RavenJObject>()).ToArray();
				var duplicate = documents.FirstOrDefault(document => jsonDocuments.Any(x => x.Key == document.Key));
				if (duplicate != null)
				{
					throw new InvalidOperationException("Found document with id: " + duplicate.Key + " on more than a single shard, which is not allowed. Document keys have to be unique cluster-wide.");
				}

				jsonDocuments.AddRange(documents);
			}

			Result = jsonDocuments
				.Select(sessionOperations.TrackEntity<T>)
				.ToArray();
		}

        public IDisposable EnterContext()
		{
			return null;
		}

		public object ExecuteEmbedded(IDatabaseCommands commands)
		{
			return commands.StartsWith(keyPrefix, matches, start, pageSize, pagingInformation, exclude: exclude)
				.Select(sessionOperations.TrackEntity<T>)
				.ToArray();
		}

		public void HandleEmbeddedResponse(object result)
		{
			Result = result;
		}
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/client/document/batches/LazyStartsWithOperation.java
package raven.client.document.batches;

import java.util.ArrayList;
import java.util.List;

import raven.abstractions.data.GetRequest;
import raven.abstractions.data.GetResponse;
import raven.abstractions.data.JsonDocument;
import raven.abstractions.json.linq.RavenJArray;
import raven.abstractions.json.linq.RavenJObject;
import raven.client.connection.IDatabaseCommands;
import raven.client.connection.SerializationHelper;
import raven.client.document.InMemoryDocumentSessionOperations;
import raven.client.utils.UrlUtils;

public class LazyStartsWithOperation<T> implements ILazyOperation {
  private final String keyPrefix;
  private final String matches;
  private final String exclude;
  private final int start;
  private final int pageSize;
  private final InMemoryDocumentSessionOperations sessionOperations;
  private final Class<T> clazz;

  private Object result;
  private boolean requiresRetry;

  public LazyStartsWithOperation(Class<T> clazz, String keyPrefix, String matches, String exclude, int start, int pageSize, InMemoryDocumentSessionOperations sessionOperations) {
    this.clazz = clazz;
    this.keyPrefix = keyPrefix;
    this.matches = matches;
    this.exclude = exclude;
    this.start = start;
    this.pageSize = pageSize;
    this.sessionOperations = sessionOperations;
  }

  @Override
  public GetRequest createRequest() {
    GetRequest getRequest = new GetRequest();
    getRequest.setUrl("/docs");
    getRequest.setQuery(String.format("startsWith=%s&matches=%s&exclude=%s&start=%d&pageSize=%d", UrlUtils.escapeDataString(keyPrefix),
        UrlUtils.escapeDataString(matches != null ? matches : ""),
        UrlUtils.escapeDataString(exclude != null ? exclude : ""),
        start, pageSize));
    return getRequest;
  }

  @Override
  public Object getResult() {
    return result;
  }

  public void setResult(Object result) {
    this.result = result;
  }

  @Override
  public boolean isRequiresRetry() {
    return requiresRetry;
  }

  public void setRequiresRetry(boolean requiresRetry) {
    this.requiresRetry = requiresRetry;
  }

  @Override
  public void handleResponse(GetResponse response) {
    if (response.isRequestHasErrors()) {
      result = null;
      requiresRetry = false;
      return;
    }

    List<JsonDocument> jsonDocuments = SerializationHelper.ravenJObjectsToJsonDocuments(((RavenJArray)response.getResult()).values(RavenJObject.class));

    List<Object> resultList = new ArrayList<>();
    for (JsonDocument doc: jsonDocuments) {
      resultList.add(sessionOperations.trackEntity(clazz, doc));
    }
    result = resultList.toArray();
  }

  @Override
  public AutoCloseable enterContext() {
    return null;
  }

  public Object executeEmbedded(IDatabaseCommands commands) {
    List<JsonDocument> docs = commands.startsWith(keyPrefix, matches, start, pageSize);
    List<Object> entites = new ArrayList<>();
    for (JsonDocument doc: docs) {
      entites.add(sessionOperations.trackEntity(clazz, doc));
    }
    return entites.toArray();
  }

  public void handleEmbeddedResponse(Object result) {
    this.result = result;
  }


}
